[Lens] Improve error message for scripted (and runtime) fields

dej611 opened the pull request:
## Summary

Fixes #95965

In Lens there's no clear visual distinction between runtime and scripted fields, really.
Also the error message retrieved from ES is kind of generic under this aspect, so I thought to make the error explain that it's a scripted field error rather than `runtime` or else.

Not really sure about the layout of the message either, probably some multiline may be helpful?

<img width="588" alt="Screenshot 2021-04-08 at 19 11 46" src="https://user-images.githubusercontent.com/924948/114069285-1eddb380-989f-11eb-9305-67c37e8adca5.png">

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/)
* Commit: 54fba4eb7522bff701fc5f7650be7dd0093047c3
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-96611/54fba4eb7522bff701fc5f7650be7dd0093047c3)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/execution/node/189/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/","number":"118331","commit":"54fba4eb7522bff701fc5f7650be7dd0093047c3"}]}
PIPELINE-->

dej611 commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118625/)
* Commit: 8409dc6185f15fb8450702ed219bfad8c82b728f
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-96611/8409dc6185f15fb8450702ed219bfad8c82b728f)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/51ba5f106dfa3f68d355ed5f45cff9c20e4bdab6) | [after](https://github.com/elastic/kibana/commit/8409dc6185f15fb8450702ed219bfad8c82b728f) | diff |
| --- | --- | --- | --- |
| `lens` | 941.7KB | 942.1KB | +455.0B |

### History
* :broken_heart: [Build #118331](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/) failed 54fba4eb7522bff701fc5f7650be7dd0093047c3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/","number":"118331","commit":"54fba4eb7522bff701fc5f7650be7dd0093047c3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118625/","number":"118625","commit":"8409dc6185f15fb8450702ed219bfad8c82b728f"}]}
PIPELINE-->

elasticmachine commented:
Pinging @elastic/kibana-app (Team:KibanaApp)

flash1293 reviewed (commented):
Can we leverage the `caused_by` object? It seems like that's the actual interesting bit. I used your example (number runtime field trying to parse a string at run time) and it doesn't tell you the actual error (just parrotting the script)

<img width="704" alt="Screenshot 2021-04-13 at 10 12 06" src="https://user-images.githubusercontent.com/1508364/114519886-273c4280-9c41-11eb-85fd-cfb6d070009a.png">
<img width="503" alt="Screenshot 2021-04-13 at 10 12 03" src="https://user-images.githubusercontent.com/1508364/114519904-2acfc980-9c41-11eb-9b0b-738a800d3812.png">

dej611 commented:
Revisited the message:

<img width="487" alt="Screenshot 2021-04-13 at 14 48 11" src="https://user-images.githubusercontent.com/924948/114555172-bb210500-9c67-11eb-8178-b7403400daff.png">

While testing it I found that having the script it's useful to have a bit more context on the error. Also, I've revisited the breaking rule for messages as the current setting wasn't great IMHO. Set new breaking rule [at word level](https://elastic.github.io/eui/#/utilities/css-utility-classes)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/119194/)
* Commit: 442b38512eac5d5f526b375569ec697cfa79e9c8
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-96611/442b38512eac5d5f526b375569ec697cfa79e9c8)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/51ba5f106dfa3f68d355ed5f45cff9c20e4bdab6) | [after](https://github.com/elastic/kibana/commit/442b38512eac5d5f526b375569ec697cfa79e9c8) | diff |
| --- | --- | --- | --- |
| `lens` | 941.7KB | 942.5KB | +859.0B |

### History
* :green_heart: [Build #118625](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118625/) succeeded 8409dc6185f15fb8450702ed219bfad8c82b728f
* :broken_heart: [Build #118331](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/) failed 54fba4eb7522bff701fc5f7650be7dd0093047c3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/","number":"118331","commit":"54fba4eb7522bff701fc5f7650be7dd0093047c3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118625/","number":"118625","commit":"8409dc6185f15fb8450702ed219bfad8c82b728f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/119194/","number":"119194","commit":"442b38512eac5d5f526b375569ec697cfa79e9c8"}]}
PIPELINE-->

wylieconlon commented on x-pack/plugins/lens/public/editor_frame_service/error_helper.ts:
```diff
@@ -26,22 +27,37 @@ const isRequestError = (e: Error | RequestError): e is RequestError => {
+  e: ElasticsearchErrorClause
+): { type: string; reason: string; nestedError?: { type: string; reason: string } } {
+  if (isScriptedFieldError(e)) {
+    const nestedError = e.failed_shards![0].reason;
```
I think you can't just look at the first shard for the script error. Here is a contrived example that I haven't tested:

* You have an index pattern with mapping conflicts, maybe one field is a keyword and one is a full text field. Full text fields aren't scriptable in aggs.
* Your script will fail on the indices that have the full text, but not the indices with keywords.

dej611 replied:
I think this is partially true. I'll try to collect them all, but also in the `data` plugin all errors but the first one are discarded.

wylieconlon commented on x-pack/plugins/lens/public/editor_frame_service/error_helper.ts:
```diff
@@ -13,6 +13,7 @@ interface ElasticsearchErrorClause {
   type: string;
   reason: string;
   caused_by?: ElasticsearchErrorClause;
+  failed_shards?: Array<{ reason: ElasticsearchErrorClause & { script?: string; lang?: string } }>;
```
It looks like this isn't typed as part of the `@elastic/elasticsearch` native types that we offer. Maybe we can use `data/common/search/search_source/types.ts` instead of writing our own here? It would also be nice if we could get error types from the elasticsearch-js client. cc @delvedor

delvedor replied:
Currently, we are shipping a type with the generic structure of Elasticsearch errors, there is an issue (https://github.com/elastic/elastic-client-generator/issues/340) about providing more advanced types, but it's not our top priority right now.

wylieconlon reviewed (commented):
I haven't tested this code yet, but my meta comment is that I'd really like to use shared logic to do error handling. It's way easier to maintain error handling in one place.

flash1293 commented:
I agree with @wylieconlon here - it's dangerous to just check the first shard, there might be different errors in different shards.

But showing errors from all shards could lead to a lot of duplicated information (e.g. if a script is broken in some way which is the main case we want to solve here). Is there an elegant way to de-dupe the errors?

I don't think we need to deliver 100% perfect error reporting for 7.13, the most important thing that should work is to report a runtime script error in a nice way to make sure runtime field integration is usable.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120070/)
* Commit: dea91fe03b8150d020b36ed148176e4ccb5df728
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check i18n](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120070/execution/node/438/log)

### History
* :green_heart: [Build #119194](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/119194/) succeeded 442b38512eac5d5f526b375569ec697cfa79e9c8
* :green_heart: [Build #118625](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118625/) succeeded 8409dc6185f15fb8450702ed219bfad8c82b728f
* :broken_heart: [Build #118331](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/) failed 54fba4eb7522bff701fc5f7650be7dd0093047c3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/","number":"118331","commit":"54fba4eb7522bff701fc5f7650be7dd0093047c3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118625/","number":"118625","commit":"8409dc6185f15fb8450702ed219bfad8c82b728f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/119194/","number":"119194","commit":"442b38512eac5d5f526b375569ec697cfa79e9c8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120070/","number":"120070","commit":"dea91fe03b8150d020b36ed148176e4ccb5df728"}]}
PIPELINE-->

dej611 commented:
I want down few routes here and refactored the code:

* I've checked with the new ES type definition of the es client, but they were too far away from this scope
* I've tried to use the `data` plugin definitions and their utilites, but they were not quite mapping the exact shape of the error

At the end I've decided to stick with the `data` plugin shape, just remodelling them a bit to better match the error coming from the `expression`, and at the same time try to have something close to the new ES type signatures.

There's currently no exact matching between what in the `data` plugin type and the new ES signatures. Surprisingly there are few shapes not mapped yet in the new signatures (will probably report a different issue in their repo).

Also, all errors are now collected from all the shards and presented - hence the change of the function name/signature.

All the functions around painless errors in the `data` plugin are focused on helping the user fixing some syntax error, which is not really helpful here. Therefore I've tried to generalized two types of errors: one type with simple `type` and `reason` strings, and another more complex for scripts where there's also some `context` object carried around, and shown.
The current `context` is only managed for `painless` scripts for now, but that may be extended in the future for other types of errors. Also the `reason` in the paniless error is really not much helpful, so I'm using it only as fallback if `script` is not provided.

<img width="686" alt="Screenshot 2021-04-15 at 18 55 52" src="https://user-images.githubusercontent.com/924948/114913841-f404d980-9e21-11eb-81fc-50d68d47cd5b.png">

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120092/)
* Commit: ddeace73ca40fa9a8294c09dee326589f2b31e42
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-96611/ddeace73ca40fa9a8294c09dee326589f2b31e42)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/83831ce99081da056c2dac59333267b56622b884) | [after](https://github.com/elastic/kibana/commit/ddeace73ca40fa9a8294c09dee326589f2b31e42) | diff |
| --- | --- | --- | --- |
| `lens` | 943.7KB | 944.6KB | +950.0B |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/83831ce99081da056c2dac59333267b56622b884) | [after](https://github.com/elastic/kibana/commit/ddeace73ca40fa9a8294c09dee326589f2b31e42) | diff |
| --- | --- | --- | --- |
| `data` | 788.7KB | 789.0KB | +222.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/83831ce99081da056c2dac59333267b56622b884) | [after](https://github.com/elastic/kibana/commit/ddeace73ca40fa9a8294c09dee326589f2b31e42) | diff |
| --- | --- | --- | --- |
| `data` | 3436 | 3447 | +11 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/83831ce99081da056c2dac59333267b56622b884) | [after](https://github.com/elastic/kibana/commit/ddeace73ca40fa9a8294c09dee326589f2b31e42) | diff |
| --- | --- | --- | --- |
| `data` | 2944 | 2954 | +10 |

#### API count with any type
| id | [before](https://github.com/elastic/kibana/commit/83831ce99081da056c2dac59333267b56622b884) | [after](https://github.com/elastic/kibana/commit/ddeace73ca40fa9a8294c09dee326589f2b31e42) | diff |
| --- | --- | --- | --- |
| `data` | 81 | 82 | +1 |

#### Non-exported public API item count
| id | [before](https://github.com/elastic/kibana/commit/83831ce99081da056c2dac59333267b56622b884) | [after](https://github.com/elastic/kibana/commit/ddeace73ca40fa9a8294c09dee326589f2b31e42) | diff |
| --- | --- | --- | --- |
| `data` | 76 | 75 | -1 |


</details>

### History
* :broken_heart: [Build #120070](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120070/) failed dea91fe03b8150d020b36ed148176e4ccb5df728
* :green_heart: [Build #119194](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/119194/) succeeded 442b38512eac5d5f526b375569ec697cfa79e9c8
* :green_heart: [Build #118625](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118625/) succeeded 8409dc6185f15fb8450702ed219bfad8c82b728f
* :broken_heart: [Build #118331](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/) failed 54fba4eb7522bff701fc5f7650be7dd0093047c3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/","number":"118331","commit":"54fba4eb7522bff701fc5f7650be7dd0093047c3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118625/","number":"118625","commit":"8409dc6185f15fb8450702ed219bfad8c82b728f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/119194/","number":"119194","commit":"442b38512eac5d5f526b375569ec697cfa79e9c8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120070/","number":"120070","commit":"dea91fe03b8150d020b36ed148176e4ccb5df728"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120092/","number":"120092","commit":"ddeace73ca40fa9a8294c09dee326589f2b31e42"}]}
PIPELINE-->

mbondyra commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120448/)
* Commit: 6376f2d2453a04d870171afb973f55212428b906
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-96611/6376f2d2453a04d870171afb973f55212428b906)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/1bc7e5462f821feff24f0470d5921d456670f646) | [after](https://github.com/elastic/kibana/commit/6376f2d2453a04d870171afb973f55212428b906) | diff |
| --- | --- | --- | --- |
| `lens` | 943.8KB | 944.7KB | +950.0B |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/1bc7e5462f821feff24f0470d5921d456670f646) | [after](https://github.com/elastic/kibana/commit/6376f2d2453a04d870171afb973f55212428b906) | diff |
| --- | --- | --- | --- |
| `data` | 789.8KB | 790.0KB | +222.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/1bc7e5462f821feff24f0470d5921d456670f646) | [after](https://github.com/elastic/kibana/commit/6376f2d2453a04d870171afb973f55212428b906) | diff |
| --- | --- | --- | --- |
| `data` | 3440 | 3451 | +11 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/1bc7e5462f821feff24f0470d5921d456670f646) | [after](https://github.com/elastic/kibana/commit/6376f2d2453a04d870171afb973f55212428b906) | diff |
| --- | --- | --- | --- |
| `data` | 2945 | 2955 | +10 |

#### API count with any type
| id | [before](https://github.com/elastic/kibana/commit/1bc7e5462f821feff24f0470d5921d456670f646) | [after](https://github.com/elastic/kibana/commit/6376f2d2453a04d870171afb973f55212428b906) | diff |
| --- | --- | --- | --- |
| `data` | 79 | 80 | +1 |

#### Non-exported public API item count
| id | [before](https://github.com/elastic/kibana/commit/1bc7e5462f821feff24f0470d5921d456670f646) | [after](https://github.com/elastic/kibana/commit/6376f2d2453a04d870171afb973f55212428b906) | diff |
| --- | --- | --- | --- |
| `data` | 76 | 75 | -1 |


</details>

### History
* :green_heart: [Build #120092](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120092/) succeeded ddeace73ca40fa9a8294c09dee326589f2b31e42
* :broken_heart: [Build #120070](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120070/) failed dea91fe03b8150d020b36ed148176e4ccb5df728
* :green_heart: [Build #119194](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/119194/) succeeded 442b38512eac5d5f526b375569ec697cfa79e9c8
* :green_heart: [Build #118625](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118625/) succeeded 8409dc6185f15fb8450702ed219bfad8c82b728f
* :broken_heart: [Build #118331](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/) failed 54fba4eb7522bff701fc5f7650be7dd0093047c3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118331/","number":"118331","commit":"54fba4eb7522bff701fc5f7650be7dd0093047c3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118625/","number":"118625","commit":"8409dc6185f15fb8450702ed219bfad8c82b728f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/119194/","number":"119194","commit":"442b38512eac5d5f526b375569ec697cfa79e9c8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120070/","number":"120070","commit":"dea91fe03b8150d020b36ed148176e4ccb5df728"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120092/","number":"120092","commit":"ddeace73ca40fa9a8294c09dee326589f2b31e42"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120448/","number":"120448","commit":"6376f2d2453a04d870171afb973f55212428b906"}]}
PIPELINE-->

wylieconlon reviewed (commented):
I haven't tested this again since my last review, but I'm not sure I know all the cases that you've already tried. Maybe unit tests would help to enumerate all the errors that you've seen?

flash1293 commented on x-pack/plugins/lens/public/editor_frame_service/error_helper.ts:
```diff
@@ -26,54 +30,82 @@ const isRequestError = (e: Error | RequestError): e is RequestError => {
+  }
+  const [payload] = getNestedErrorClause(causedBy);
+  if (lang === 'painless') {
+    return [{ ...payload, context: { type: 'Painless script', reason: `"${script}"` || reason } }];
```
should we i18n this? Seems like kind of a grey area...

dej611 replied:
Thought about i18n here, but it would be the only message localized. The original string reports something like `"script error"` or similar, I'm just being a bit more precise.

flash1293 reviewed (commented):
Tested around a bit and runtime field errors are handled great now:
<img width="571" alt="Screenshot 2021-04-20 at 11 05 55" src="https://user-images.githubusercontent.com/1508364/115369462-6a119380-a1c8-11eb-8d46-147a99909e7e.png">

Agree about unit tests for the different cases we expect to handle - it will make it much easier to generalize/extend this later on.

dej611 commented:
> Tested around a bit and runtime field errors are handled great now:
> <img alt="Screenshot 2021-04-20 at 11 05 55" width="571" src="https://user-images.githubusercontent.com/1508364/115369462-6a119380-a1c8-11eb-8d46-147a99909e7e.png">
> 
> Agree about unit tests for the different cases we expect to handle - it will make it much easier to generalize/extend this later on.

TBH I feel that a distinct flow for runtime errors would be even better, given the amount of information retrieved probably a different layout of the message would be preferable.
The current state is ok, I guess, in terms of content, less in terms of presentation.

flash1293 commented:
I agree that a special format for those would be even better, but I don't want to bloat the scope of this too much because I would like to get it in 7.13 (which feature-freezes today)

dej611 commented:
I've attempted to reproduce various types of errors (also based on previous #79831 history) but since then few things have changed, like error messages and error handling (some of the things that could be broken at the time I could not break any more).

For instance `scripted field`s are no longer reported as `Painless` script type of errors, so I've added some check for that. in the test file there's the whole error passed from ES for it to check.

So I've managed to reduce error to only few "families" and add some testing for them.
Also, added one more runtime field handling when user could access the indexpattern saved object but could not access the actual indexes underneath: that was triggering an internal esaggs error which wasn't bubbling up.

<img width="668" alt="Screenshot 2021-04-20 at 14 37 58" src="https://user-images.githubusercontent.com/924948/115398818-d51e9280-a1e7-11eb-9e3f-3f1c7a45e4d6.png">

flash1293 commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120935/)
* Commit: 5e011f9df5d7edbb7cc209eaa13104e39497fe63
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-96611/5e011f9df5d7edbb7cc209eaa13104e39497fe63)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/1bc7e5462f821feff24f0470d5921d456670f646) | [after](https://github.com/elastic/kibana/commit/5e011f9df5d7edbb7cc209eaa13104e39497fe63) | diff |
| --- | --- | --- | --- |
| `lens` | 943.8KB | 945.1KB | +1.3KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/1bc7e5462f821feff24f0470d5921d456670f646) | [after](https://github.com/elastic/kibana/commit/5e011f9df5d7edbb7cc209eaa13104e39497fe63) | diff |
| --- | --- | --- | --- |
| `data` | 789.8KB | 790.0KB | +222.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/1bc7e5462f821feff24f0470d5921d456670f646) | [after](https://github.com/elastic/kibana/commit/5e011f9df5d7edbb7cc209eaa13104e39497fe63) | diff |
| --- | --- | --- | --- |
| `data` | 3440 | 3451 | +11 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/1bc7e5462f821feff24f0470d5921d456670f646) | [after](https://github.com/elastic/kibana/commit/5e011f9df5d7edbb7cc209eaa13104e39497fe63) | diff |
| --- | --- | --- | --- |
| `data` | 2945 | 2955 | +10 |

#### API count with any type
| id | [before](https://github.com/elastic/kibana/commit/1bc7e5462f821feff24f0470d5921d456670f646) | [after](https://github.com/elastic/kibana/commit/5e011f9df5d7edbb7cc209eaa13104e39497fe63) | diff |
| --- | --- | --- | --- |
| `data` | 79 | 80 | +1 |

#### Non-exported public API item count
| id | [before](https://github.com/elastic/kibana/commit/1bc7e5462f821feff24f0470d5921d456670f646) | [after](https://github.com/elastic/kibana/commit/5e011f9df5d7edbb7cc209eaa13104e39497fe63) | diff |
| --- | --- | --- | --- |
| `data` | 76 | 75 | -1 |


</details>

### History
* :green_heart: [Build #120448](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120448/) succeeded 6376f2d2453a04d870171afb973f55212428b906
* :green_heart: [Build #120092](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120092/) succeeded ddeace73ca40fa9a8294c09dee326589f2b31e42
* :broken_heart: [Build #120070](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120070/) failed dea91fe03b8150d020b36ed148176e4ccb5df728
* :green_heart: [Build #119194](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/119194/) succeeded 442b38512eac5d5f526b375569ec697cfa79e9c8
* :green_heart: [Build #118625](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118625/) succeeded 8409dc6185f15fb8450702ed219bfad8c82b728f

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/118625/","number":"118625","commit":"8409dc6185f15fb8450702ed219bfad8c82b728f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/119194/","number":"119194","commit":"442b38512eac5d5f526b375569ec697cfa79e9c8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120070/","number":"120070","commit":"dea91fe03b8150d020b36ed148176e4ccb5df728"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120092/","number":"120092","commit":"ddeace73ca40fa9a8294c09dee326589f2b31e42"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120448/","number":"120448","commit":"6376f2d2453a04d870171afb973f55212428b906"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120935/","number":"120935","commit":"5e011f9df5d7edbb7cc209eaa13104e39497fe63"}]}
PIPELINE-->

flash1293 reviewed (approved):
I tested the following cases:
* Runtime field error ✅ 
* Scripted field error ✅ 
* No access to source indices - in most cases fails on the index pattern level because the fields are not available which is misleading (but not a failure of this PR - we should tackle it separately)
* Too much data: "Batch request failed with 413" (not too bad, searching for 413 gives a good hint)

Looks good to me for 7.13, but we open an issue for a more general solution to handling these errors.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120974/)
* Commit: c03aae1171b1df73eebcf496d4ac5fb4298da6b9
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-96611/c03aae1171b1df73eebcf496d4ac5fb4298da6b9)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/948aa3a9f556a29a9523914c29518d32c43e1fb8) | [after](https://github.com/elastic/kibana/commit/c03aae1171b1df73eebcf496d4ac5fb4298da6b9) | diff |
| --- | --- | --- | --- |
| `lens` | 947.0KB | 948.3KB | +1.3KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/948aa3a9f556a29a9523914c29518d32c43e1fb8) | [after](https://github.com/elastic/kibana/commit/c03aae1171b1df73eebcf496d4ac5fb4298da6b9) | diff |
| --- | --- | --- | --- |
| `data` | 789.8KB | 790.0KB | +222.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/948aa3a9f556a29a9523914c29518d32c43e1fb8) | [after](https://github.com/elastic/kibana/commit/c03aae1171b1df73eebcf496d4ac5fb4298da6b9) | diff |
| --- | --- | --- | --- |
| `data` | 3443 | 3454 | +11 |

#### API count missing comments
| id | [before](https://github.com/elastic/kibana/commit/948aa3a9f556a29a9523914c29518d32c43e1fb8) | [after](https://github.com/elastic/kibana/commit/c03aae1171b1df73eebcf496d4ac5fb4298da6b9) | diff |
| --- | --- | --- | --- |
| `data` | 2948 | 2958 | +10 |

#### API count with any type
| id | [before](https://github.com/elastic/kibana/commit/948aa3a9f556a29a9523914c29518d32c43e1fb8) | [after](https://github.com/elastic/kibana/commit/c03aae1171b1df73eebcf496d4ac5fb4298da6b9) | diff |
| --- | --- | --- | --- |
| `data` | 79 | 80 | +1 |

#### Non-exported public API item count
| id | [before](https://github.com/elastic/kibana/commit/948aa3a9f556a29a9523914c29518d32c43e1fb8) | [after](https://github.com/elastic/kibana/commit/c03aae1171b1df73eebcf496d4ac5fb4298da6b9) | diff |
| --- | --- | --- | --- |
| `data` | 76 | 75 | -1 |


</details>

### History
* :green_heart: [Build #120935](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120935/) succeeded 5e011f9df5d7edbb7cc209eaa13104e39497fe63
* :green_heart: [Build #120448](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120448/) succeeded 6376f2d2453a04d870171afb973f55212428b906
* :green_heart: [Build #120092](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120092/) succeeded ddeace73ca40fa9a8294c09dee326589f2b31e42
* :broken_heart: [Build #120070](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120070/) failed dea91fe03b8150d020b36ed148176e4ccb5df728
* :green_heart: [Build #119194](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/119194/) succeeded 442b38512eac5d5f526b375569ec697cfa79e9c8

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/119194/","number":"119194","commit":"442b38512eac5d5f526b375569ec697cfa79e9c8"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120070/","number":"120070","commit":"dea91fe03b8150d020b36ed148176e4ccb5df728"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120092/","number":"120092","commit":"ddeace73ca40fa9a8294c09dee326589f2b31e42"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120448/","number":"120448","commit":"6376f2d2453a04d870171afb973f55212428b906"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120935/","number":"120935","commit":"5e011f9df5d7edbb7cc209eaa13104e39497fe63"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/120974/","number":"120974","commit":"c03aae1171b1df73eebcf496d4ac5fb4298da6b9"}]}
PIPELINE-->

tsullivan reviewed (approved):
app-services changes LGTM

kibanamachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [7.x](https://github.com/elastic/kibana/pull/97785)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/97785">](https://github.com/elastic/kibana/pull/97785) |

This backport PR will be merged automatically after passing CI.
